exchanges/okcoin/okcoin.go

```diff
@@ -12,613 +12,1658 @@ import (
-	resp := make(map[string][]PlaceOrderResponse)
-	for i := range request {
-		currencyPairOrders[request[i].InstrumentID]++
+// --------------------- Trade endpoints ---------------------------
+
+func (a *PlaceTradeOrderParam) validateTradeOrderParameter() error {
+	if a == nil {
+		return errNilArgument
 	}
-	if len(currencyPairOrders) > 4 {
-		return resp, []error{errors.New("up to 4 trading pairs")}
+	if a.InstrumentID.IsEmpty() {
+		return fmt.Errorf("%w, instrument id is required", currency.ErrCurrencyPairEmpty)
 	}
-	for _, orderCount := range currencyPairOrders {
-		if orderCount > 4 {
-			return resp, []error{errors.New("maximum 4 orders for each pair")}
-		}
+	if a.TradeMode == "" {
+		return errTradeModeIsRequired
 	}
+	if a.Side == "" {
+		return fmt.Errorf("%w, empty order side", order.ErrSideIsInvalid)
+	}
+	if a.OrderType == "" {
+		return fmt.Errorf("%w, empty order type", order.ErrTypeIsInvalid)
+	}
+	if a.Size <= 0 {
+		return fmt.Errorf("%w, size: %f", errInvalidAmount, a.Size)
+	}
+	return nil
+}
 
-	err := o.SendHTTPRequest(ctx, exchange.RestSpot, http.MethodPost, marginTradingSubsection, batchOrders, request, &resp, true)
+// PlaceOrder to place a trade order.
+func (o *Okcoin) PlaceOrder(ctx context.Context, arg *PlaceTradeOrderParam) (*TradeOrderResponse, error) {
+	err := arg.validateTradeOrderParameter()
+	if err != nil {
+		return nil, err
+	}
+	var resp []TradeOrderResponse
+	err = o.SendHTTPRequest(ctx, exchange.RestSpot, placeTradeOrderEPL, http.MethodPost, typeTrade, "order", arg, &resp, true)
 	if err != nil {
-		return resp, []error{err}
+		return nil, err
+	}
+	if len(resp) == 0 {
+		return nil, errNoValidResponseFromServer
 	}
+	if resp[0].SCode != "0" {
+		return nil, fmt.Errorf("code: %s msg: %s", resp[0].SCode, resp[0].SMsg)
+	}
+	return &resp[0], nil
+}
 
-	var orderErrors []error
-	for currency, orderResponse := range resp {
-		for i := range orderResponse {
-			if !orderResponse[i].Result {
-				orderErrors = append(orderErrors, fmt.Errorf("order for currency %v failed to be placed", currency))
-			}
+// PlaceMultipleOrder place orders in batches. Maximum 20 orders can be placed per request. Request parameters should be passed in the form of an array.
+func (o *Okcoin) PlaceMultipleOrder(ctx context.Context, args []PlaceTradeOrderParam) ([]TradeOrderResponse, error) {
+	var err error
+	if len(args) == 0 {
+		return nil, fmt.Errorf("%w, 0 length place order requests", errNilArgument)
+	}
+	for x := range args {
+		err = args[x].validateTradeOrderParameter()
+		if err != nil {
+			return nil, err
 		}
 	}
+	var resp []TradeOrderResponse
+	return resp, o.SendHTTPRequest(ctx, exchange.RestSpot, placeTradeMultipleOrdersEPL, http.MethodPost, typeTrade, "batch-orders", &args, &resp, true)
+}
 
-	return resp, orderErrors
+// CancelTradeOrder cancels a single trade order
+func (o *Okcoin) CancelTradeOrder(ctx context.Context, arg *CancelTradeOrderRequest) (*TradeOrderResponse, error) {
```

Comment:
orderID and clientOtderID are optional, your work here is correct, just add omitempty to their fields in the request